lasagne/layers/recurrent.py

```diff
@@ -0,0 +1,1353 @@
+            Symbolic output variable.
+        """
+        # Retrieve the layer input
+        input = inputs[0]
+        # Retrieve the mask when it is supplied
+        mask = inputs[1] if len(inputs) > 1 else None
+
+        # Treat all dimensions after the second as flattened feature dimensions
+        if input.ndim > 3:
+            input = input.reshape((input.shape[0], input.shape[1],
+                                   T.prod(input.shape[2:])))
+
+        # Input should be provided as (n_batch, n_time_steps, n_features)
+        # but scan requires the iterable dimension to be first
+        # So, we need to dimshuffle to (n_time_steps, n_batch, n_features)
+        input = input.dimshuffle(1, 0, 2)
+        seq_len, num_batch, _ = input.shape
+
+        if self.precompute_input:
+            # Because the input is given for all time steps, we can precompute
+            # the inputs to hidden before scanning. First we need to reshape
+            # from (seq_len, batch_size, num_inputs) to
+            # (seq_len*batch_size, num_inputs)
+            input = T.reshape(input,
+                              (seq_len*num_batch, -1))
+            input = helper.get_output(
+                self.input_to_hidden, input, **kwargs)
+
+            # Reshape back to (seq_len, batch_size, num_units)
+            input = T.reshape(input, (seq_len, num_batch, -1))
+
+        # We will always pass the hidden-to-hidden layer params to step
+        non_seqs = helper.get_all_params(self.hidden_to_hidden)
+        # When we are not precomputing the input, we also need to pass the
+        # input-to-hidden parameters to step
+        if not self.precompute_input:
+            non_seqs += helper.get_all_params(self.input_to_hidden)
+
+        # Create single recurrent computation step function
+        def step(input_n, hid_previous, *args):
+            # Compute the hidden-to-hidden activation
+            hid_pre = helper.get_output(self.hidden_to_hidden, hid_previous)
+
+            # If the dot product is precomputed then add it, otherwise
+            # calculate the input_to_hidden values and add them
+            if self.precompute_input:
+                hid_pre += input_n
+            else:
+                hid_pre += helper.get_output(self.input_to_hidden, input_n)
+
+            # Clip gradients
+            if self.grad_clipping is not False:
+                hid_pre = theano.gradient.grad_clip(
+                    hid_pre, -self.grad_clipping, self.grad_clipping)
+
+            return self.nonlinearity(hid_pre)
+
+        def step_masked(input_n, mask_n, hid_previous, *args):
+            # If mask is 0, use previous state until mask = 1 is found.
+            # This propagates the layer initial state when moving backwards
+            # until the end of the sequence is found.
+            hid = step(input_n, hid_previous, *args)
+            hid_out = hid*mask_n + hid_previous*(1 - mask_n)
+            return [hid_out]
```

Comment:
So actually you can do a lot of weird and wonderful things with masks, like skipping some mid-sequence inputs, or mixing the hidden state with the previous state. Maybe the comment should just be:
```python
        # Skip over any input with mask 0 by copying the previous hidden state,
        # proceed normally for any input with mask 1.
```
The current comment focuses on moving backwards, which I find confusing because it applies to other cases as well, depending on how the mask is set up. (Although I know that you have in mind a certain type of mask that is also described in the docstring.)